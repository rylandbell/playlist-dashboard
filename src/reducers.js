import { combineReducers } from 'redux';

const accessToken = (state = "", action) => {
  switch (action.type){
    case 'SET_ACCESS_TOKEN':
      return action.data;
    default:
      return state;
  }
}

const playlists = (state = [], action) => {
  switch (action.type){
    case 'ADD_PLAYLISTS_DATA':
      return action.data.slice();
    case 'TOGGLE_PLAYLIST_SELECT':
      const playlists = state.slice();
      const chosen = playlists.find(playlist => playlist.id === action.id);
      chosen.selected ? chosen.selected = false : chosen.selected = true;
      return playlists;
    default:
      return state;
  }
}

const receivedPlaylists = (state = false, action) => {
  switch(action.type) {
    case 'ADD_PLAYLISTS_DATA':
      return true;
    default:
      return state;
  }
}

const activeView = (state = "preAuth", action) => {
  switch (action.type){
    case 'CHANGE_VIEW':
      return action.view;
    case 'SET_ACCESS_TOKEN':
      const isToken = action.data && action.data.length > 0;
      const view = isToken ? "selectPlaylists" : "preAuth";
      return view;
    default:
      return state;
  }
}

const app = combineReducers({
  accessToken,
  playlists,
  activeView,
  receivedPlaylists
});

export default app;
