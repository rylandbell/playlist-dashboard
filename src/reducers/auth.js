export const accessToken = (state = "", action) => {
  switch (action.type){
    case 'SET_ACCESS_TOKEN':
      return action.data;
    default:
      return state;
  }
}

export const authStatus = (state = false, action) => {
  switch (action.type){
    case 'SET_ACCESS_TOKEN':
      return action.data && action.data.length > 0;
    default:
      return state;
  }
}

export const badAuthToken = (state = false, action) => {
  switch (action.type){
    case 'BAD_AUTH_TOKEN':
      return true;
    default:
      return state;
  }
}

export const userId = (state = "", action) => {
  switch (action.type){
    case 'ADD_PLAYLISTS_DATA':
      const href = action.payload.href;
      const userId = href.split('/')[5];
      return userId;
    default:
      return state;
  }
}