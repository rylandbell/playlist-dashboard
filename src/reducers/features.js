const initialFeatureValues = [
  {
    name: 'danceability',
    isActive: true,
    isGraphed: true,
    isDim: false,
    sortBy: false,
    showReferenceLine: false,
    currentValue: [0, 1]
  },
  {
    name: 'energy',
    isActive: true,
    isGraphed: false,
    isDim: false,
    sortBy: false,
    showReferenceLine: false,
    currentValue: [0, 1]
  },
  {
    name: 'valence',
    isActive: true,
    isGraphed: false,
    isDim: false,
    sortBy: false,
    showReferenceLine: false,
    currentValue: [0, 1]
  },
  {
    name: 'popularity',
    isActive: true,
    isGraphed: false,
    isDim: false,
    sortBy: false,
    showReferenceLine: false,
    currentValue: [0, 1]
  },
  {
    name: 'acousticness',
    isActive: true,
    isGraphed: false,
    isDim: false,
    sortBy: false,
    showReferenceLine: false,
    currentValue: [0, 1]
  },
  {
    name: 'instrumentalness',
    isActive: true,
    isGraphed: false,
    isDim: false,
    sortBy: false,
    showReferenceLine: false,
    currentValue: [0, 1]
  }
  // {
  //   name: 'liveness',
  //   isActive: true,
  //   isGraphed: false,
  // isDim: false,
  //   showReferenceLine: false,
  //   currentValue: [0,1],
  // },
];

export const features = (state = initialFeatureValues, action) => {
  let targetFeature, updatedFeature;
  switch (action.type) {
    case 'UPDATE_FEATURE_FILTER':
      targetFeature = state[action.featureIndex];
      updatedFeature = { ...targetFeature, currentValue: action.data };
      return [
        ...state.slice(0, action.featureIndex),
        updatedFeature,
        ...state.slice(action.featureIndex + 1)
      ];
    case 'TOGGLE_CHARTED_FEATURE':
      return state.map((feature, index) => {
        if (index === action.featureIndex) {
          return { ...feature, isGraphed: action.newValue };
        } else {
          return feature;
        }
      });
    case 'START_DRAGGING_FEATURE_SLIDER':
      return state.map((feature, index) => {
        if (index === action.featureIndex) {
          return { ...feature, showReferenceLine: true, isGraphed: true };
        } else {
          return { ...feature, isDim: true };
        }
      });
    case 'STOP_DRAGGING_FEATURE_SLIDER':
      return state.map(feature => ({
        ...feature,
        showReferenceLine: false,
        isDim: false
      }));
    case 'SORT_BY_FEATURE':
      return state.map(feature => {
        if (feature.name === action.payload.featureName) {
          
          // cycle through 'ascending', 'descending', false:
          let newSortByValue
          if (feature.sortBy === 'ascending') {
            newSortByValue = 'descending';
          } else if (feature.sortBy === 'descending') {
            newSortByValue = false;
          } else {
            newSortByValue = 'ascending';
          }

          return {...feature, sortBy: newSortByValue};

        //set all other features to unsorted:
        } else {
          return {...feature, sortBy: false};
        }
      });
    default:
      return state;
  }
};
