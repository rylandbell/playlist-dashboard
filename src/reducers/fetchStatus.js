export const getPlaylistsPending = (state = false, action) => {
  switch (action.type) {
    case 'GET_PLAYLISTS_PENDING':
      return true;
    case 'GET_PLAYLISTS_SUCCESS':
      return false;
    case 'GET_PLAYLISTS_FAILURE':
      return false;
    default:
      return state;
  }
};

export const getPlaylistsSuccess = (state = false, action) => {
  switch (action.type) {
    case 'GET_PLAYLISTS_PENDING':
      return false;
    case 'GET_PLAYLISTS_SUCCESS':
      return true;
    case 'GET_PLAYLISTS_FAILURE':
      return false;
    default:
      return state;
  }
};

export const getPlaylistsFailure = (state = false, action) => {
  switch (action.type) {
    case 'GET_PLAYLISTS_PENDING':
      return false;
    case 'GET_PLAYLISTS_SUCCESS':
      return false;
    case 'GET_PLAYLISTS_FAILURE':
      return true;
    default:
      return state;
  }
};

export const getTracksPending = (state = false, action) => {
  switch (action.type) {
    case 'GET_TRACKS_PENDING':
      return true;
    case 'GET_TRACKS_SUCCESS':
      return false;
    case 'GET_TRACKS_FAILURE':
      return false;
    default:
      return state;
  }
};

export const getTracksSuccess = (state = false, action) => {
  switch (action.type) {
    case 'GET_TRACKS_PENDING':
      return false;
    case 'GET_TRACKS_SUCCESS':
      return true;
    case 'GET_TRACKS_FAILURE':
      return false;
    default:
      return state;
  }
};

export const getTracksFailure = (state = false, action) => {
  switch (action.type) {
    case 'GET_TRACKS_PENDING':
      return false;
    case 'GET_TRACKS_SUCCESS':
      return false;
    case 'GET_TRACKS_FAILURE':
      return true;
    default:
      return state;
  }
};

export const getFeaturesPending = (state = false, action) => {
  switch (action.type) {
    case 'GET_FEATURES_PENDING':
      return true;
    case 'GET_FEATURES_SUCCESS':
      return false;
    case 'GET_FEATURES_FAILURE':
      return false;
    default:
      return state;
  }
};

export const getFeaturesSuccess = (state = false, action) => {
  switch (action.type) {
    case 'GET_FEATURES_PENDING':
      return false;
    case 'GET_FEATURES_SUCCESS':
      return true;
    case 'GET_FEATURES_FAILURE':
      return false;
    default:
      return state;
  }
};

export const getFeaturesFailure = (state = false, action) => {
  switch (action.type) {
    case 'GET_FEATURES_PENDING':
      return false;
    case 'GET_FEATURES_SUCCESS':
      return false;
    case 'GET_FEATURES_FAILURE':
      return true;
    default:
      return state;
  }
};

export const createPlaylistPending = (state = false, action) => {
  switch (action.type) {
    case 'CREATE_PLAYLIST_PENDING':
      return true;
    case 'CREATE_PLAYLIST_SUCCESS':
      return false;
    case 'CREATE_PLAYLIST_FAILURE':
      return false;
    default:
      return state;
  }
};

export const createPlaylistSuccess = (state = false, action) => {
  switch (action.type) {
    case 'CREATE_PLAYLIST_PENDING':
      return false;
    case 'CREATE_PLAYLIST_SUCCESS':
      return true;
    case 'CREATE_PLAYLIST_FAILURE':
      return false;
    default:
      return state;
  }
};

export const createPlaylistFailure = (state = false, action) => {
  switch (action.type) {
    case 'CREATE_PLAYLIST_PENDING':
      return false;
    case 'CREATE_PLAYLIST_SUCCESS':
      return false;
    case 'CREATE_PLAYLIST_FAILURE':
      return true;
    default:
      return state;
  }
};

export const addTracksToPlaylistPending = (state = false, action) => {
  switch (action.type) {
    case 'ADD_TRACKS_TO_PLAYLIST_PENDING':
      return true;
    case 'ADD_TRACKS_TO_PLAYLIST_SUCCESS':
      return false;
    case 'ADD_TRACKS_TO_PLAYLIST_FAILURE':
      return false;
    default:
      return state;
  }
};

export const addTracksToPlaylistSuccess = (state = false, action) => {
  switch (action.type) {
    case 'ADD_TRACKS_TO_PLAYLIST_PENDING':
      return false;
    case 'ADD_TRACKS_TO_PLAYLIST_SUCCESS':
      return true;
    case 'ADD_TRACKS_TO_PLAYLIST_FAILURE':
      return false;
    default:
      return state;
  }
};

export const addTracksToPlaylistFailure = (state = false, action) => {
  switch (action.type) {
    case 'ADD_TRACKS_TO_PLAYLIST_PENDING':
      return false;
    case 'ADD_TRACKS_TO_PLAYLIST_SUCCESS':
      return false;
    case 'ADD_TRACKS_TO_PLAYLIST_FAILURE':
      return true;
    default:
      return state;
  }
};
