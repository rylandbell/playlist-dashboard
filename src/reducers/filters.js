const filtersInitial = [
  {
    name: 'danceability',
    isActive: true,
    isGraphed: true,
    isDim: false,
    showReferenceLine: false,
    currentValue: [0, 1]
  },
  {
    name: 'energy',
    isActive: true,
    isGraphed: false,
    isDim: false,
    showReferenceLine: false,
    currentValue: [0, 1]
  },
  {
    name: 'valence',
    isActive: true,
    isGraphed: false,
    isDim: false,
    showReferenceLine: false,
    currentValue: [0, 1]
  },
  {
    name: 'popularity',
    isActive: true,
    isGraphed: false,
    isDim: false,
    showReferenceLine: false,
    currentValue: [0, 1]
  },
  {
    name: 'acousticness',
    isActive: true,
    isGraphed: false,
    isDim: false,
    showReferenceLine: false,
    currentValue: [0, 1]
  },
  {
    name: 'instrumentalness',
    isActive: true,
    isGraphed: false,
    isDim: false,
    showReferenceLine: false,
    currentValue: [0, 1]
  }
  // {
  //   name: 'liveness',
  //   isActive: true,
  //   isGraphed: false,
  // isDim: false,
  //   showReferenceLine: false,
  //   currentValue: [0,1],
  // },
];

export const filterValues = (state = filtersInitial, action) => {
  let targetFilter, updatedFilter;
  switch (action.type) {
    case 'UPDATE_FILTER':
      targetFilter = state[action.filterIndex];
      updatedFilter = { ...targetFilter, currentValue: action.data };
      return [
        ...state.slice(0, action.filterIndex),
        updatedFilter,
        ...state.slice(action.filterIndex + 1)
      ];
    case 'TOGGLE_CHARTED_FEATURE':
      return state.map((filter, index) => {
        if (index === action.filterIndex) {
          return { ...filter, isGraphed: action.newValue };
        } else {
          return filter;
        }
      });
    case 'START_DRAGGING_FEATURE_SLIDER':
      return state.map((filter, index) => {
        if (index === action.filterIndex) {
          return { ...filter, showReferenceLine: true, isGraphed: true };
        } else {
          return { ...filter, isDim: true };
        }
      });
    case 'STOP_DRAGGING_FEATURE_SLIDER':
      return state.map(filter => ({
        ...filter,
        showReferenceLine: false,
        isDim: false
      }));
    case 'SELECT_PLAYLIST':
      return state.map(filter => ({ ...filter, showReferenceLine: false }));
    default:
      return state;
  }
};
