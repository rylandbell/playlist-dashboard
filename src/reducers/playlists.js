export const playlists = (state = [], action) => {
  switch (action.type){
    case 'ADD_PLAYLISTS_DATA':
      return action.payload.items
        .map(playlist => ({
          id: playlist.id,
          name: playlist.name,
          ownerId: playlist.owner.id,
          images: playlist.images
        }));
    default:
      return state;
  }
}

export const selectedPlaylist = (state = {}, action) => {
  switch(action.type) {
    case 'SELECT_PLAYLIST':
      return {...action.data};
    default:
      return state;
  }
}