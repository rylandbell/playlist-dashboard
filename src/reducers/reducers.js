import { combineReducers } from 'redux';
import * as fetchStatusReducers from './fetchStatus';

export const fetchStatus = combineReducers(fetchStatusReducers);

export const accessToken = (state = "", action) => {
  switch (action.type){
    case 'SET_ACCESS_TOKEN':
      return action.data;
    default:
      return state;
  }
}

export const authStatus = (state = "preAuth", action) => {
  switch (action.type){
    case 'SET_ACCESS_TOKEN':
      return action.data && action.data.length > 0;
    default:
      return state;
  }
}

export const userId = (state = "", action) => {
  switch (action.type){
    case 'ADD_PLAYLISTS_DATA':
      const href = action.data.href;
      const userId = href.split('/')[5];
      return userId;
    default:
      return state;
  }
}

export const playlists = (state = [], action) => {
  switch (action.type){
    case 'ADD_PLAYLISTS_DATA':
      return action.data.items.slice();
    default:
      return state;
  }
}

export const selectedPlaylist = (state = null, action) => {
  switch(action.type) {
    case 'SELECT_PLAYLIST':
      return action.data;
    default:
      return state;
  }
}

export const selectedPlaylistTracks = (state = [], action) => {
  switch (action.type){
    case 'SELECT_PLAYLIST':
      return action.data ? [] : state;
    case 'ADD_TRACKS_DATA':
      return action.data.slice();
    default:
      return state;
  }
}

export const selectedPlaylistAudioFeatures = (state = [], action) => {
  switch (action.type){
    case 'SELECT_PLAYLIST':
      return action.data ? [] : state;
    case 'ADD_AUDIO_FEATURES':
      return action.data.slice();
    default:
      return state;
  }
}

export const hoveredTrackRow = (state = null, action) => {
  switch (action.type){
    case 'HOVER_ON_TRACK':
      return action.data;
    case 'CLEAR_HOVERED_TRACK':
      return null
    default:
      return state;
  }
}

export const activeSidebarTab = (state = "playlists", action) => {
  switch (action.type) {
    case 'SET_ACTIVE_TAB':
      return action.data;
    case 'SELECT_PLAYLIST':
      return action.forceTabSwitch ? "filters" : state;
    default:
      return state;
  }
}

// on first playlist select, automatically push user to Filters tab. after that, let user control tabs
export const autoSidebarTabSwitch = (state = true, action) => {
  switch (action.type) {
    case 'SELECT_PLAYLIST':
      return false;
    default:
      return state;
  }
}

const filtersInitial = [
  {
    name: 'danceability',
    displayName: 'Danceability',
    shortName: 'Dance',
    isActive: true,
    isGraphed: true,
    showReferenceLine: false,
    currentValue: [0,1],
    min: 0,
    max: 1,
    color: '#5bc0de'
  },
  {
    name: 'energy',
    displayName: 'Energy',
    shortName: 'Energy',
    isActive: true,
    isGraphed: false,
    showReferenceLine: false,
    currentValue: [0,1],
    min: 0,
    max: 1,
    color: '#d9534f'
  },
  {
    name: 'valence',
    displayName: 'Positivity',
    shortName: 'Pos',
    isActive: true,
    isGraphed: false,
    showReferenceLine: false,
    currentValue: [0,1],
    min: 0,
    max: 1,
    color: '#5CB85C'
  },
  {
    name: 'acousticness',
    displayName: 'Acousticness',
    shortName: 'Acous',
    isActive: true,
    isGraphed: false,
    showReferenceLine: false,
    currentValue: [0,1],
    min: 0,
    max: 1,
    color: '#f0ad4e'
  },
  {
    name: 'liveness',
    displayName: 'Liveness',
    shortName: 'Live',
    isActive: true,
    isGraphed: false,
    showReferenceLine: false,
    currentValue: [0,1],
    min: 0,
    max: 1,
    color: '#DF691A'
  },
  {
    name: 'instrumentalness',
    displayName: 'Instrumentalness',
    shortName: 'Inst',
    isActive: true,
    isGraphed: false,
    showReferenceLine: false,
    currentValue: [0,1],
    min: 0,
    max: 1,
    color: '#EBEBEB'
  }
];

export const filters = (state = filtersInitial, action) => {
  let targetFilter, updatedFilter;
  switch(action.type) {
    case 'UPDATE_FILTER':
      targetFilter = Object.assign({}, state[action.filterIndex]);
      updatedFilter = Object.assign(targetFilter, {currentValue: action.data});
      return [
        ...state.slice(0, action.filterIndex),
        updatedFilter,
        ...state.slice(action.filterIndex + 1)
      ];
    case 'TOGGLE_CHARTED_FEATURE':
      return state.map( (filter, index) => {
        if (index === action.filterIndex) {
          return Object.assign(filter, {showReferenceLine: false, isGraphed: action.newValue})
        } else {
          return Object.assign({}, filter, {showReferenceLine: false});
        }
      });
    case 'START_DRAGGING_FEATURE_SLIDER':
      return state.map( (filter,index) => {
        if (index === action.filterIndex) {
          return Object.assign({}, filter, {showReferenceLine: true, isGraphed: true});
        } else {
          return Object.assign({}, filter, {showReferenceLine: false, isGraphed: false});
        }
      });
    case 'SELECT_PLAYLIST':
      return state.map( filter => Object.assign(filter, {showReferenceLine: false}) );
    default:
      return state;
  }
}

export const newPlaylistName = (state = "", action) => {
  switch (action.type) {
    case 'SELECT_PLAYLIST':
      return "Modified: " + action.data.name;
    case 'CHANGE_NAME_TEXT':
      return action.data;
    default:
      return state;
  }
}

export const animateNextChartDraw = (state = true, action) => {
  switch (action.type) {
    case 'STOP_ANIMATING_CHART':
      return false;
    case 'ADD_AUDIO_FEATURES':
      return true;
    default:
      return state;
  }
}


