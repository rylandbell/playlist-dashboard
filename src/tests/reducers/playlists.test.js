import deepFreeze from 'deep-freeze';
import * as reducers from '../../reducers/playlists';

describe("Playlist reducers", () => {

  describe('playlists', () => {
    const initState = reducers.playlists(undefined, {type: 'INIT'});

    it('should return the data of an ADD_PLAYLIST_DATA action', () => {
      const action = {
        type: 'ADD_PLAYLISTS_DATA', 
        payload: {
          items: [
            {
              id: 'testId',
              name: 'testName',
              owner: {
                id: 'testOwnerId'
              },
              fake: 'fake',
              images: []
            },
            {
              id: 'testId2',
              name: 'testName2',
              owner: {
                id: 'testOwnerId2'
              },
              fake: 'fake',
              images: []
            }
          ]
        }
      };

      deepFreeze(action);
      deepFreeze(initState);
      expect(reducers.playlists(initState, action)).toMatchSnapshot();
    });
  });

  describe('selectedPlaylist', () => {
    const initState = reducers.selectedPlaylist(undefined, {type: 'INIT'});

    it('should return the data of a SELECT_PLAYLIST action', () => {
      const action = {type: 'SELECT_PLAYLIST', data: {id: 'test'}};
      deepFreeze(action);
      deepFreeze(initState);
      expect(reducers.selectedPlaylist(initState, action)).toMatchSnapshot();
    });
  });

});