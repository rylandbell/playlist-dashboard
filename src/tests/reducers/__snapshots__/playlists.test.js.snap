// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Playlist reducers playlists should return the data of an ADD_PLAYLIST_DATA action 1`] = `
Array [
  Object {
    "id": "testId",
    "images": Array [],
    "name": "testName",
    "ownerId": "testOwnerId",
  },
  Object {
    "id": "testId2",
    "images": Array [],
    "name": "testName2",
    "ownerId": "testOwnerId2",
  },
]
`;

exports[`Playlist reducers selectedPlaylist should return the data of a SELECT_PLAYLIST action 1`] = `
Object {
  "id": "test",
}
`;
